lib/tpm2_kdfe.c

```diff
@@ -0,0 +1,238 @@
+        LOG_ERR("Could not convert EC public key to BN");
+        goto out;
+    }
+    q = EC_POINT_new(group);
+    if(q == NULL) {
+        LOG_ERR("Could not allocate EC_POINT");
+        goto out;
+    }
+
+    if(0 == EC_POINT_set_affine_coordinates_GFp(group, q, bn_Qx, bn_Qy, NULL)) {
+        LOG_ERR("Could not set affine_coordinates");
+        EC_POINT_free(q);
+        q = NULL;
+        goto out;
+    }
+out:
+    if(bn_Qx) BN_free(bn_Qx);
+    if(bn_Qy) BN_free(bn_Qy);
+    if(key) EC_KEY_free(key);
+
+    return q;
+}
+
+
+static bool get_public_key_from_ec_key(EC_KEY *key, TPMS_ECC_POINT *point) {
+    BIGNUM *x = BN_new();
+    BIGNUM *y = BN_new();
+    const EC_POINT *pubkey = EC_KEY_get0_public_key(key);
+    int nbx, nby;
+    bool result = false;
+
+    if((x == NULL) || (y == NULL) || (pubkey == NULL)) {
+        LOG_ERR("Failed to allocate memory to store EC public key.");
+        goto out;
+    }
+
+    EC_POINT_get_affine_coordinates_GFp(EC_KEY_get0_group(key),
+            pubkey, x, y, NULL);
+    nbx =  BN_num_bytes(x);
+    nby =  BN_num_bytes(y);
+    if( (nbx > sizeof(point->x.buffer))||
+            (nby > sizeof(point->y.buffer))) {
+        LOG_ERR("EC public key has too many bits.");
+        goto out;
+    }
+
+    point->x.size = nbx;
+    point->y.size = nby;
+    BN_bn2bin(x, point->x.buffer);
+    BN_bn2bin(y, point->y.buffer);
+    result = true;
+
+out:
+    if(x) BN_free(x);
+    if(y) BN_free(y);
+    return result;
+}
+
+
+static int get_ECDH_shared_secret(
+        EC_KEY *key,
+        const EC_POINT *pPub,
+        TPM2B_ECC_PARAMETER *secret) {
+
+    int sharedSecretLength;
+
+    sharedSecretLength = EC_GROUP_get_degree(EC_KEY_get0_group(key));
+    sharedSecretLength = (sharedSecretLength + 7) / 8;
+    if(sharedSecretLength > sizeof(secret->buffer)) {
+        return -1;
+    }
+    secret->size = ECDH_compute_key(secret->buffer, sharedSecretLength, pPub, key, NULL);
+    return secret->size;
+}
+
+
+
+bool ecdh_derive_seed_and_encrypted_seed(
+        TPM2B_PUBLIC *parent_pub,
+        unsigned char *label, int labelLen,
+        TPM2B_DIGEST *seed,
+        TPM2B_ENCRYPTED_SECRET *outSymSeed) {
+
+    TPMS_ECC_PARMS *tpm_ecc = &parent_pub->publicArea.parameters.eccDetail;
+    TPMI_ALG_HASH parent_name_alg = parent_pub->publicArea.nameAlg;
+    UINT16 parent_hash_size = tpm2_alg_util_get_hash_size(parent_name_alg);
+    bool result = false;
+    EC_KEY *key = NULL;
+    EC_POINT *Qsv = NULL;
+    TPMS_ECC_POINT Qeu;
+    TPM2B_ECC_PARAMETER eccSecret;
+
+    // generate an ephemeral key
+    int nid = tpm2_ossl_curve_to_nid(tpm_ecc->curveID);
+    if (nid >= 0) {
+        key = EC_KEY_new_by_curve_name(nid);
+    }
+    if(key == NULL) {
+        LOG_ERR("Failed to create EC key from curveID");
+        return false;
+    }
+    EC_KEY_generate_key(key);
+
+    // get public key for the ephemeral key
+    if(false == get_public_key_from_ec_key(key, &Qeu)) {
```

Comment:
space after if.

Reply:
Do not hide the function in the conditional statement, assign the function return to a boolean and use that boolean in the conditional statement.